### PR TITLE
feat: precision metric for classification

### LINKED ISSUE #185: Precision metric for classification
### Is your feature request related to a problem?

Computing the precision of a classifier for a specific class is a common task.

### Desired solution

* New method called `precision` in `Classifier`
* Required parameter `positive_class` to specify which value should be treated as positive (all other classes are treated as negative)
* Call [`sklearn.metrics.precision`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html#sklearn.metrics.precision_score)

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by PhilipGutberlet**
Closes #185.

### Summary of Changes
Added precision function in _classifier.

Co-authored-by: 1288832338+patrikguempel@users.noreply.github.com

### CODE REVIEW

**Comment by lars-reimann** [comment]
## [🦙 MegaLinter](https://megalinter.io/6.22.2) status: ❌ [ERROR](https://github.com/Safe-DS/Stdlib/actions/runs/4892421070)

| Descriptor  |                                 Linter                                 |Files|Fixed|                            Errors                            |Elapsed time|
|-------------|------------------------------------------------------------------------|-----|----:|--------------------------------------------------------------|------------|
|✅ PYTHON    |[black](https://megalinter.io/6.22.2/descriptors/python_black)          |    2|    1|                                                             0|0.75s       |
|❌ PYTHON    |[mypy](https://megalinter.io/6.22.2/descriptors/python_mypy)            |    2|     |[1](https://github.com/Safe-DS/Stdlib/actions/runs/4892421070)|0.25s       |
|✅ PYTHON    |[ruff](https://megalinter.io/6.22.2/descriptors/python_ruff)            |    2|    1|                                                             0|0.05s       |
|✅ REPOSITORY|[git_diff](https://megalinter.io/6.22.2/descriptors/repository_git_diff)|yes  |     |no                                                            |0.03s       |

See detailed report in [MegaLinter reports](https://github.com/Safe-DS/Stdlib/actions/runs/4892421070)
_Set `VALIDATE_ALL_CODEBASE: true` in mega-linter.yml to validate all sources, not only the diff_

_MegaLinter is graciously provided by [OX Security](https://www.ox.security/?ref=megalinter)_

File: tests/safeds/ml/classical/classification/test_classifier.py [hunk 1]
```diff
@@ -234,4 +266,4 @@ def test_with_different_types(self) -> None:
     )
     def test_should_raise_if_table_is_not_tagged(self, table: Table) -> None:
         with pytest.raises(UntaggedTableError):
-            DummyClassifier().accuracy(table)  # type: ignore[arg-type]
```
Comment (lars-reimann) [inline tests/safeds/ml/classical/classification/test_classifier.py:237]: Should not be changed.

File: src/safeds/ml/classical/classification/_classifier.py [hunk 2]
```diff
@@ -98,3 +99,39 @@ def accuracy(self, validation_or_test_set: TaggedTable) -> float:
         predicted = self.predict(validation_or_test_set.features).target
 
         return sk_accuracy_score(expected._data, predicted._data)
+
+    def precision(self, validation_or_test_set: TaggedTable, positive_class=1) -> float:
+        """
+        Compute the classifier's precision of the on the given data.
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:105]:
```suggestion
        Compute the classifier's precision on the given data.
```

File: src/safeds/ml/classical/classification/_classifier.py [hunk 3]
```diff
@@ -98,3 +99,39 @@ def accuracy(self, validation_or_test_set: TaggedTable) -> float:
...
+
+        Parameters
+        ----------
+        validation_or_test_set : TaggedTable
+            The validation or test set.
+        positive_class : int | str
+            The classification to be considered positive
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:112]:
```suggestion
            The class to be considered positive. All other classes are considered negative.
```

File: src/safeds/ml/classical/classification/_classifier.py [hunk 4]
```diff
@@ -98,3 +99,39 @@ def accuracy(self, validation_or_test_set: TaggedTable) -> float:
...
+        -------
+        precision : float
+            The calculated precision score, i.e. the ratio of correctly predicted positives to all predicted positives.
+            Returns 1 if no predictions are made.
+        """
+        expected = validation_or_test_set.target
+        predicted = self.predict(validation_or_test_set.features).target
+
+        if len(expected) != len(predicted):
+            raise AssertionError("Different length of 'expected' and 'predicted' vectors.")
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:124]: Not covered in test. Use a `ColumnLengthMismatchError` instead of an `AssertionError`.

File: src/safeds/ml/classical/classification/_classifier.py [hunk 5]
```diff
@@ -98,3 +99,39 @@ def accuracy(self, validation_or_test_set: TaggedTable) -> float:
...
+
+        true_positive, false_positive = 0, 0
+
+        for i in range(len(expected)):
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:128]: Have a look at the `enumerate` or `zip` built-in functions.

See: https://stackoverflow.com/a/35775980

File: src/safeds/ml/classical/classification/_classifier.py [hunk 6]
```diff
@@ -98,3 +99,39 @@ def accuracy(self, validation_or_test_set: TaggedTable) -> float:
...
+            if predicted[i] == positive_class:
+                if expected[i] == predicted[i]:
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:130]:
```suggestion
                if expected[i] == positive_class:
```

It's equivalent but I think it better conveys the intention that the positive class is expected.

File: tests/safeds/ml/classical/classification/test_classifier.py [hunk 7]
```diff
@@ -219,6 +219,38 @@ def test_with_different_types(self) -> None:
...
+        table = Table.from_dict(
+            {
+                "predicted": [1, "1", "0", "2"],
+                "expected": [1, 0, 1, 2],
+            },
+        ).tag_columns(target_name="expected")
+
+        assert DummyClassifier().precision(table, 1) == 1.0
+
+    def test_should_return_1(self) -> None:
```
Comment (lars-reimann) [inline tests/safeds/ml/classical/classification/test_classifier.py:244]:
```suggestion
    def test_should_return_1_if_never_expected_to_be_positive(self) -> None:
```

**Review by lars-reimann — CHANGES_REQUESTED** [review]

**Review by lars-reimann — APPROVED** [review]
